Add Hausdorff Distance trait

JosiahParry opened the pull request:
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/main/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.

---

This PR adds a new trait `HausdorffDistance` which is implemented for all geometries. 

This is my first Rust library PR so I'll need some guidance getting this through the finish line! 

### Outstanding questions for maintainers: 

- This PR is lacking tests. Should there be a test for each combo of geometry?
- `CHANGES.md` does not have a development version listed, how should it be modified?
- what amount of documentation is needed / suggested?

frewsxcv commented on geo/src/algorithm/hausdorff_distance.rs:
```diff
@@ -0,0 +1,194 @@
+
+/// Determine the distance between two geometries using the [Hausdorff distance formula].
+///
+/// [haversine formula]: https://en.wikipedia.org/wiki/Hausdorff_distance
```
> `[haversine formula]`

This should be updated

frewsxcv commented on geo/src/algorithm/hausdorff_distance.rs:
```diff
@@ -0,0 +1,194 @@
+                            geom
+                            .coords_iter()
+                            .map(|c2| {
+                                c.euclidean_distance(&c2)
```
Can you clarify in the docs that this algorithm utilizes euclidean distance under the hood? In case someone might be expecting it to use haversine or something else

JosiahParry commented:
@frewsxcv thanks! I've added a note in the documentation that it uses euclidean distance. Is this explicit enough?

https://github.com/JosiahParry/geo/blob/1c44e3b5578b4e2af24bf3755827ca2b87c7e2eb/geo/src/algorithm/hausdorff_distance.rs#L9-L10

I've also added a handful of tests as well. Please let me know if this is sufficient or if there ought to be exhaustive tests (every combo of geometry types).

JosiahParry commented on geo/src/algorithm/hausdorff_distance.rs:
```diff
@@ -0,0 +1,278 @@
+                            .map(|c2| {
+                                c.euclidean_distance(&c2)
+                            })
+                            .fold(<T as Bounded>::max_value(), |accum, val| accum.min(val))
```
Erm, I don't know? Since the map returns `T` not `Option<T>` does `.unwrap_or()` have a use here beyond calling the or portion? Additionally, I was unaware you can call `min()` on an iterator!

JosiahParry replied:
I've tried replacing this with `.min()` and received the following compiler error. It appears that since `Ord` trait is not implemented for `T` `min()` cannot be used. 
<img width="604" alt="image" src="https://github.com/georust/geo/assets/13534508/2790fbe1-ea9d-4390-8f65-6fb8b3869d9b">

JosiahParry commented:
I've pushed a commit that addresses formatting using `cargo fmt`.

michaelkirk commented on geo/src/algorithm/hausdorff_distance.rs:
```diff
@@ -0,0 +1,261 @@
+    };
+}
+
+impl_hausdorff_distance_coord_iter! {
```
Thanks for getting comprehensive implementations with the macros. 

I think you could delete a lot of this code if you implemented generically, based on the `CoordsIter` trait.

The impl's would look something like this:

```rust
impl<'a, 'b, T, G1, G2> HausdorffDistance<'a, 'b, T, G2> for G1
where
    T: GeoFloat,
    G1: CoordsIter<'a, Scalar = T>,
    G2: CoordsIter<'b, Scalar = T>,
{
    fn hausdorff_distance(&'a self, rhs: &'b G2) -> T {
       todo!()
    }
}

impl<'a, 'b, T, G> HausdorffDistance<'a, 'b, T, G> for Coord<T>
where
    T: GeoFloat,
    G: CoordsIter<'b, Scalar = T>
{
    fn hausdorff_distance(&'a self, geom: &'b G) -> T {
        todo()!
    }
}
```

... and you'd have to add the lifetimes to the trait declaration. We've done similar things for ConvexHull.

What do you think?

JosiahParry replied:
I've been trying to get this sorted out for a little while this morning but I'm finding myself in lifetime hell. It seems like there might be a mismatch between lifetimes between `CoordsIter` and `T`. 

I'll check the implementation for ConvexHull and see if i can get it sorted out.

JosiahParry replied:
Resolved in newest commit.

michaelkirk commented on geo/src/algorithm/mod.rs:
```diff
@@ -120,6 +120,10 @@ pub use geodesic_intermediate::GeodesicIntermediate;
 pub mod geodesic_length;
 pub use geodesic_length::GeodesicLength;
 
+/// Calculate the Hausdorff distance between two geometries.
```
Thank you for maintaining alphabetical order in this file. ❤️

JosiahParry commented:
I've added a note to the changelog. Please let me know what else is needed to close this PR. Thanks! :)

michaelkirk commented:
bors r+

michaelkirk commented:
bors retry

Whoops. My bad I think. Let's try again now that I have (for real this time) disabled the merge queue.

michaelkirk commented:
Thanks!

JosiahParry commented:
Thank you!
